lib/SPIRV/libSPIRV/SPIRVModule.cpp

```diff
@@ -1069,15 +1078,70 @@ SPIRVValue *SPIRVModuleImpl::addNullConstant(SPIRVType *Ty) {
 
 SPIRVValue *SPIRVModuleImpl::addCompositeConstant(
     SPIRVType *Ty, const std::vector<SPIRVValue *> &Elements) {
+  const size_t MaxNumElements = MaxWordCount - SPIRVConstantComposite::FixedWC;
+  const size_t NumElements = Elements.size();
+  if (NumElements > MaxNumElements &&
+      isAllowedToUseExtension(ExtensionID::SPV_INTEL_long_constant_composite)) {
+    SPIRVConstantComposite *Res;
+    for (uint64_t J = 0; J < NumElements; J += MaxNumElements) {
+      auto Start = Elements.begin() + J;
+      auto End = ((J + MaxNumElements) > NumElements) ? Elements.end()
+                                                      : Start + MaxNumElements;
+      std::vector<SPIRVValue *> Slice(Start, End);
+      if (J == 0) {
+        Res = static_cast<SPIRVConstantComposite *>(
+            addCompositeConstant(Ty, Slice));
+      } else {
+        SPIRVConstantComposite::ContinuedInstType Continued =
+            static_cast<SPIRVConstantComposite::ContinuedInstType>(
+                addCompositeConstantContinuedINTEL(Slice));
+        Res->addContinuedInstruction(Continued);
+      }
+    }
+    return Res;
```

Comment:
Ok, maybe this variant?
```suggestion
    auto Start = Elements.begin();
    auto End = Start + MaxNumElements;
    std::vector<SPIRVValue *> Slice(Start, End);
    auto *Res = static_cast<SPIRVConstantComposite *>(addCompositeConstant(Ty, Slice));
    for (; End != Elements.end();) {
      Start = End;
      End = ((Elements.end() - End) > MaxNumElements) ? End + MaxNumElements
                                                      : Elements.end();
      Slice.assign(Start, End);
      SPIRVConstantComposite::ContinuedInstType Continued =
          static_cast<SPIRVConstantComposite::ContinuedInstType>(
              addCompositeConstantContinuedINTEL(Slice));
      Res->addContinuedInstruction(Continued);
    }
    return Res;
```

Reply:
Just to:
* remove extra check for J == 0 on each iteration;
* use looping via iterators instead of index

Reply:
Ok, I'll give it a try.